### ISSUE TITLE
ExamReview #4

### ISSUE THREAD

**Issue by rayshroom** [seneca-btp500f19/btp500-f19#19]
Why is Mergesort O(nlogn)?

I found it is super hard to explain with English only.

A:
[Assume this is a image, where [ ] is a element]
Divide stage: log n steps in total
[    ][    ][    ][    ][    ][    ][    ][    ]
[   ][   ][   ][   ]        [   ][   ][   ][   ]    constant steps
[  ][  ]    [  ][  ]        [  ][  ]    [  ][  ]    constant steps
[ ]  [ ]    [ ]  [ ]        [ ]  [ ]    [ ]  [ ]    constant steps
Merge stage: log n steps in total, where each takes n/2 comparison
[  ][  ]    [  ][  ]        [  ][  ]    [  ][  ]    n/2 (comparison + swap)
[   ][   ][   ][   ]        [   ][   ][   ][   ]    n/2 (comparison + swap)
[    ][    ][    ][    ][    ][    ][    ][    ]    n/2 (comparison + swap)

Let T(n) be the total number of steps to perform a mergesort.
Since mergesort is composed of 2 stages, a divide stage and a merge stage.
Let D(n) be the total number of steps to divide the input(array) down to minimal units(elements)
Let M(n) be the total number of steps to merge the minimal units(elements) back to the input(array).
Therefore T(n) = D(n) + M(n).
D(n) = c * log n, where c represents the constant number of steps taken by division per step, with log n steps in total
M(n) = (n/2 + d) * log n, where each step takes n/2 number of constant comparison with swapping, and there are log n steps in total
T(n) = D(n) + M(n) + c * log n + n/2 * log n = (c + n/2) * log n
Therefore T(n) is O(n log n)

**Comment by saketepe** [comment]
Wouldn't mergesort be nlogn because of the second half of the sort? do you even need to mention the first half?

**Comment by rayshroom** [comment]
Yes, its mainly the second half, first half is put in for the analysis to be more "complete"
